### DISCUSSION TITLE
Implicit conversions for interface properties

### DISCUSSION THREAD

**Discussion by CrosRoad95** [dotnet/csharplang#6226] [Language Ideas]
Let interfaces declare implicit properties, when delcared all implementations have to inherit from that type
For example:
```csharp
public interface IFoo
{
  public implicit IEnumerable<int> Numbers { get; }
}

public class Foo : IFoo
{
  public List<int> Numbers { get; set; }
}

public class Bar : IFoo
{
  public HashSet<int> Numbers { get; set; }
}
```

now you have to declare separete property `public IEnumerable<int> Numbers => NumbersList;`

**Comment by VWilcox2000** [comment]
I like this idea.  Could also be considered for acting a a modifier on the interface that would do the same thing -- make it so all properties can be implemented with super-classes that are implicitly castable to the interface item.

I think your use case is the no-brainer easy change , another scenario I've disliked since they introduced generics after ArrayList being pretty much all you had was the inability for things like IComparable<IFoo> to implicitly support IComparable.   I think it's less important now as most library support the generic, but still sometimes you need both as like so:

      public interface IFoo : IComparable<IFoo>, IComparable
      {
      }

      public class Foo : IFoo
      {
        public int CompareTo(IFoo? other)
        {
          ...
        }

        public int CompareTo(object? obj)
        {
          ...
        }
      }

CompareTo with an object could do something vastly different than CompareTo(IFoo?...) but I have never leveraged that, and on the principle of keeping code fairly easy for other programmers to understand, I doubt I would EVER have actual different purposes for the two.  Implicit keyword on IComparable could tell the compiler that CompareTo(IFoo?  being the containing class would thus work well enough for CompareTo with an implicit CompareTo(object as Foo) happening there.

      public interface IFoo : IComparable<IFoo>, implicit IComparable
      {
      }

      public class Foo : IFoo
      {
        public int CompareTo(IFoo? other)
        {
          ...
        }

But that is not a straightforward as your scenario which I think is the more important use case.
